[SR] Create and edit repositories UI

jen-huang opened the pull request:
## Summary

This PR adds ability to **create and edit repositories** from the Snapshot and Restore app UI. All repository types are supported, including plugin-based repositories! Instructions for setting up repositories are below. After creating or editing a repository, the user will be redirected back to repository list with the newly created or edited repository details flyout open.

**Note: The forms do not have client-side validation yet. I will add that next.**

This PR also adds **repository verification** information to the UI, so that users can see which repositories are connected and which ones may potentially have configuration issues.

---
### Instructions

<details>
<summary>Instructions for setting up repositories for testing</summary>

#### File system
1. Add the file system path you want to use to elasticsearch.yml or as part of starting up ES:
```
path:
  repo: /tmp/es-backups
```
or
```
yarn es snapshot --license trial -E path.repo=/tmp/es-backups
```
2. Use Console to add repository, use the file system path above as `location` setting:
```
PUT /_snapshot/my_backup
{
  "type": "fs",
  "settings": {
    "location": "/tmp/es-backups",
    "chunk_size": "10mb"
  }
}
```
3. Adjust `settings` as necessary, all available settings can be found in docs:
https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_shared_file_system_repository

#### Readonly
Readonly repositories only take `url` setting. Documentation: https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_read_only_url_repository

It's easy to set up a `file:` url:
```
PUT _snapshot/my_readonly_repository
{
  "type": "url",
  "settings": {
    "url": "file:///tmp/es-backups"
  }
}
```

### Source only
Source only repositories are special in that they are basically a wrapper around another repository type. Documentation: https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-snapshots.html#_source_only_repository

This means that the settings that are available depends on the `delegate_type` parameter. For example, this source only repository delegates to `fs` (file system) type, so all file system rules and available settings apply:
```
PUT _snapshot/my_src_only_repository
{
  "type" : "source",
  "settings" : {
    "delegate_type" : "fs",
    "location" : "/tmp/es-backups"
  }
}
```

#### Plugin-based repositories:
There are four official repository plugins available: S3, GCS, HDFS, Azure. Please refer to the meta ticket **Developer notes** section : https://github.com/elastic/elasticsearch-team/issues/118

Available plugin repository settings can be found in docs: https://www.elastic.co/guide/en/elasticsearch/plugins/master/repository.html
</details>

---

### Screenshots

Verification status in repository list table:
![image](https://user-images.githubusercontent.com/1965714/55119407-fcf5a780-50ae-11e9-8991-e25e12019c15.png)

Verification status and details in repository details:
![image](https://user-images.githubusercontent.com/1965714/55119444-1ac30c80-50af-11e9-9b6b-37dead4d5d06.png)

Register (create) repository form:
![image](https://user-images.githubusercontent.com/1965714/55119473-3dedbc00-50af-11e9-9d78-507bfe3d5320.png)

The form changes based on repository type. Here's a `source` repository type that delegates to `azure`:
![image](https://user-images.githubusercontent.com/1965714/55119553-8dcc8300-50af-11e9-9c81-37cd56427941.png)

Edit repository form, note that the `name` and `type` fields are not editable:
![image](https://user-images.githubusercontent.com/1965714/55119644-e734b200-50af-11e9-9f95-043089cb6b91.png)

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7112/)

jen-huang commented on x-pack/server/lib/create_router/error_wrappers/wrap_es_error.js:
```diff
@@ -6,6 +6,20 @@
 
 import Boom from 'boom';
 
+function extractCausedByChain(causedBy = {}, accumulator = []) {
```
the changes here are copied from the changes done in CCR to surface ES messages to UI:
https://github.com/elastic/kibana/blob/master/x-pack/plugins/cross_cluster_replication/server/lib/error_wrappers/wrap_es_error.js

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7113/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7217/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7224/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7232/)

yaronp68 commented:
@jen-huang I propose to remove the delegate combo and have a toggle for source (source only snapshots)
Another thing is to point to the [documentation of repositories plugin ](https://www.elastic.co/guide/en/elasticsearch/plugins/current/repository.html)

yaronp68 commented:
Another comment - when additional custom settings are supported instead of `{}` place holder show an example like `{my_settings: my_value}`

sebelga commented on x-pack/plugins/snapshot_restore/public/app/sections/home/home.tsx:
```diff
@@ -9,7 +9,7 @@ import { Route, RouteComponentProps, Switch } from 'react-router-dom';
 import { EuiPageBody, EuiPageContent, EuiSpacer, EuiTab, EuiTabs, EuiTitle } from '@elastic/eui';
 
-import { BASE_PATH, Section } from '../../constants';
+import { BASE_PATH, getHomeBreadcrumb, Section } from '../../constants';
```
It looks strange to me to import a method from a constants file...

To me, like we quickly discuss over slack, we should save a reference to `i18n` in some sort of `TranslateService` file, so we don't need to pass it around from the app dependencies from inside a component.

Here, if we are going to update the breadcrumb many times, I'd have a `BreadcrumbsService` with a simple api

```ts
import { breadCrumbsService } from '../services/breadcrumbs'`

breadCrumbsService.set('home'); // all the logic and translation (getHomeBreadcrumb, getRepositoryAddBreadcrumb) is done there
```

sebelga commented on x-pack/plugins/snapshot_restore/public/app/sections/home/repository_list/repository_details/repository_details.tsx:
```diff
@@ -4,51 +4,71 @@
 import {
   EuiButton,
   EuiButtonEmpty,
+  // @ts-ignore
```
Can we add the missing type in the `@elastic/eui/index.d.ts`?

sebelga commented on x-pack/plugins/snapshot_restore/public/app/sections/home/repository_list/repository_details/repository_details.tsx:
```diff
@@ -4,51 +4,71 @@
+    error,
+    loading,
+    data: { repository, verification },
+  } = useRequest({
```
To be changed if we go the "requests creator" path 😊

sebelga commented on x-pack/plugins/snapshot_restore/public/app/sections/repository_add/repository_add.tsx:
```diff
@@ -0,0 +1,102 @@
+    setIsSaving(true);
+    setSaveError(null);
+    const { name } = newRepository;
+    const { error } = await sendRequest({
```
Here too it'd be cleaner a request creator:  `await saveRepository(newRepository)`

sebelga commented on x-pack/plugins/snapshot_restore/public/app/services/utils/clean_settings.ts:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Utility to remove empty fields ("") from repository settings
+ */
+export const cleanSettings = (settings: Repository['settings']) =>
```
We might want to be able to use the same method for other forms object? What about naming it something like `removeEmptyStringsFromObject(...)`. It could even do it deeply (with an optional parameter), but then you'd be passing the interview a second time 😄

cjcenizal replied:
Just to play devil's advocate, I think there is a benefit to preserving this as specialized logic for cleaning the settings object, as opposed to generalizing it. In its current form it's really clear to me what its purpose is: empty strings have a special meaning in the context of the `settings` object, and this logic tells me that we don't want to keep them (for some reason). It's really simple to grok the intention behind this code. I understand that there's a business rule here and if we make changes to this code it will because we're adapting to changes in the business rule.

On the other hand, if we extract this out into a generalized utility, now the context has changed and we've invited new features (like the ability to clean an object deeply, which may not be necessary or even desired for our use case). If our business rules change to where we're using `null` instead of empty strings, then we'd need to either modify the generalized function or write new code.

Based on these pros and cons I really don't think we should bother generalizing this. :)

sebelga replied:
The more I keep reading about functional programming the more we will have a discrepancy on this and the way we build software. 😊 In lean towards functions that **do not** have any business "specialized logic" baked in them. I believe in small functions that do 1 thing well and that are composable (especially utilities functions).

In the **composition** of those small function resides the business logic.

In the long run, it makes code cleaner, the webpage faster (no need to send the same JS over and over in differents .bundle.js), lower memory footprint, and I would even say more robust as the same piece of code is being executed hundreds of time (and thus fully tested). 
And it goes without saying the fact that it saves in development time by not having to re-think, re-write and re-test the same exact code.

I mention it because.... I wrote a very similar function in CCR I think 😊 and the "_empty strings have a special meaning in the context of the settings object_" statement is not totally true. We need it in all those scenarios where we don't want the client to send form fields that haven't been modified.

sebelga commented on x-pack/plugins/snapshot_restore/public/plugin.ts:
```diff
@@ -27,61 +27,56 @@ export class Plugin {
 
-    const unmountReactApp = (elem: Element | undefined | null): void => {
+    const unmountReactApp = (): void => {
+      const elem = document.getElementById(REACT_ROOT_ID);
```
Are you sure about this refactor? We kept the reference to the `elem` outside so we could unmount the react app without losing the reference to the dom node (as Angular was creating a new `div` in the DOM)... 

Did you fully test this?

jen-huang replied:
yep. these changes were made due to the react app mounting twice when navigating away and coming back (React dev tools showed 2 apps)

sebelga replied:
Ok.... then we might have the bug in our other apps as they are all using this pattern. I'll check.

sebelga commented on x-pack/plugins/snapshot_restore/server/lib/booleanize_settings.ts:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Utility to coerce ES boolean strings to actual booleans
+ */
+export const booleanizeSettings = (repository: Repository) => {
```
Same comment as above. I'd give it a generic name so it could be used and reused at will 😊

cjcenizal replied:
Maybe as a compromise we can follow the rule of threes, and wait for multiple use cases to arise before generalizing?

sebelga replied:
So here I was just saying to give it a generic _name_. By now we should have a common library of all those generic utils functions instead of re-writing them over and over.

jen-huang commented on x-pack/plugins/snapshot_restore/public/app/components/repository_form/repository_form.tsx:
```diff
@@ -0,0 +1,395 @@
+    () => {
+      const repositoryTypeOptions = [...repositoryTypes];
+      const { type } = repository;
+      if (isEditing && REPOSITORY_PLUGINS_MAP[type] && !repositoryTypes.includes(type)) {
```
this is for an edge case like the following:
1. user creates plugin-based repository like `s3`
2. user uninstalls plugin for some reason
3. repository will still exist and show in the list, but the list of types available will not include `s3`, so this logic pushes the repository into the list anyway

this scenario is not encountered for creation of a new repository, hence the `isEditing` check 🙂

sebelga replied:
Ah ok, I see

jen-huang commented on x-pack/plugins/snapshot_restore/public/app/components/repository_form/type_settings/s3_settings.tsx:
```diff
@@ -0,0 +1,483 @@
+          </EuiTitle>
+        }
+        description={
+          <Fragment>
```
the perils of copy pasting from another block that _did_ need Fragment!

jen-huang commented on x-pack/plugins/snapshot_restore/public/app/components/repository_form/type_settings/azure_settings.tsx:
```diff
@@ -0,0 +1,370 @@
+    settings: { client, container, base_path, compress, chunk_size, readonly, location_mode },
+  } = repository;
+
+  const locationModeOptions = ['primary_only', 'secondary_only'].map(option => ({
```
as these refer to third-party setting values, I think localizing them would be confusing. the documentation for azure in Chinese does not translate their key names:
https://docs.microsoft.com/zh-cn/javascript/api/azure-storage/azurestorage.common.util.storageutilities.locationmode?view=azure-node-latest

sebelga replied:
Good point.

sebelga commented on x-pack/plugins/snapshot_restore/public/app/components/repository_form/repository_form.tsx:
```diff
@@ -0,0 +1,395 @@
+          renderLoadingRepositoryTypesError()
+        ) : (
+          <EuiSelect
+            disabled={isEditing}
```
I like the idea but then we'd need to update the other apps to follow that pattern, no?

jen-huang commented on x-pack/plugins/snapshot_restore/public/app/components/repository_form/type_settings/hdfs_settings.tsx:
```diff
@@ -0,0 +1,421 @@
+          <Fragment>
+            <FormattedMessage
+              id="xpack.snapshotRestore.repositoryForm.typeHDFS.uriDescription"
+              defaultMessage="The URI address for HDFS. Required."
```
in this case, most repository fields are optional rather than required, so it's less "noise" to mark required vs optional

cjcenizal replied:
Gotcha. I'll check with the design team about this later on to see if this is the approach other teams are taking too, or if we should just consistently use "optional" or "required" everywhere.

jen-huang commented on x-pack/plugins/snapshot_restore/public/app/components/repository_verification_badge.tsx:
```diff
@@ -0,0 +1,56 @@
+      <EuiHealth color="success">
+        <FormattedMessage
+          id="xpack.snapshotRestore.repositoryVerification.verificationSuccessfulValue"
+          defaultMessage="Success"
```
to me, "Not verified" means "we didn't check this repository" rather than that there was an error with connecting to the repository. although ES uses the term "verification", I think "repository status" or "connection status" is more accurate. let's defer this until copywriting review, but maybe we can simply say "Status" instead, maybe with the values "Connected" or "Error"

jen-huang commented:
Thanks for the in-depth feedback, everyone! I've addressed most of them:

* Removed `break`s
* Adjusted add and edit repo routes so they don't conflict with list route
* Added repository plugin and types doc links to form
* Bootstrap documentation service
* Bootstrap text service and replace `RepositoryTypeName` component with it
* Bootstrap breadcrumb service and replace usages
* Bootstrap httpService, remove chrome and http from app dependencies(!)
* Add request creator and replace all instances of `useRequest` and `sendRequest` with it
* Simplify update repository and update repository setting methods

The major ones that I will defer are UI changes/improvements (as suggested by @yaronp68 and @cjcenizal) as these will take longer and I'd like to get this PR merged so that the upstream feature branch is in a good place to begin shared dev work 😄

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7919/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7927/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7929/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7932/)
